HMASynthesizer throws an error when sampling multi table models with three levels of depths

### Environment Details

Please indicate the following details about the environment in which you found the bug:

* SDV version: 1.4.1.dev0
* Python version: Python 3.11.5
* Operating System: All

### Error Description

@elisaherrmann and I are getting an error when sampling a multi table model with three levels of depth and multiple root parent nodes with the HMASynthesizer.

The model we are trying to sample:

![image](https://github.com/sdv-dev/SDV/assets/2892293/d66cd2b1-fe72-4579-ad9d-749f5ed396ba)


### Steps to reproduce

We created an integration test to reproduce:

```python
   def test_hma_two_rootparents_two_children_one_grandchild(self):
            """Test it works on a 'parent-children-grandchild' dataset."""
            # Setup
            child1 = pd.DataFrame(data={
                'child_ID': ['a', 'b', 'c', 'd', 'e'],
                'parent_ID1': [0, 1, 2, 3, 3],
                'data': ['0', '1', '2', '3', '4']
            })
            child2 = pd.DataFrame(data={
                'child_ID': ['a', 'b', 'c', 'd', 'e'],
                'parent_ID2': [0, 1, 2, 3, 4],
                'data': ['0', '1', '2', '3', '4']
            })
            root_parent1 = pd.DataFrame(data={
                'parent_ID1': [0, 1, 2, 3, 4],
                'data': [True, False, False, False, True]
            })
            root_parent2 = pd.DataFrame(data={
                'parent_ID2': [0, 1, 2, 3, 4],
                'data': ['Yes', 'Yes', 'Maybe', 'No', 'No']
            })
            grandchild = pd.DataFrame(data={
                'grandchild_ID': ['a', 'b', 'c', 'd', 'e'],
                'child_ID1': ['a', 'b', 'c', 'd', 'e'],
                'child_ID2': ['a', 'b', 'c', 'd', 'e'],
                'data': ['0', '1', '2', '3', '4']
            })
            data = {'parent1': root_parent1, 'child1': child1, 'child2': child2, 'parent2': root_parent2, 'grandchild': grandchild}
            metadata = MultiTableMetadata.load_from_dict({
                'tables': {
                    'parent1': {
                        'primary_key': 'parent_ID1',
                        'columns': {
                            'parent_ID1': {'sdtype': 'id'},
                            'data': {'sdtype': 'categorical'}
                        }
                    },
                    'parent2': {
                        'primary_key': 'parent_ID2',
                        'columns': {
                            'parent_ID2': {'sdtype': 'id'},
                            'data': {'sdtype': 'categorical'}
                        }
                    },
                    'child1': {
                        'primary_key': 'child_ID',
                        'columns': {
                            'child_ID': {'sdtype': 'id'},
                            'parent_ID1': {'sdtype': 'id'},
                            'data': {'sdtype': 'categorical'}
                        }
                    },
                    'child2': {
                        'primary_key': 'child_ID',
                        'columns': {
                            'child_ID': {'sdtype': 'id'},
                            'parent_ID2': {'sdtype': 'id'},
                            'data': {'sdtype': 'categorical'}
                        }
                    },
                    'grandchild': {
                        'primary_key': 'grandchild_ID',
                        'columns': {
                            'grandchild_ID': {'sdtype': 'id'},
                            'child_ID1': {'sdtype': 'id'},
                            'child_ID2': {'sdtype': 'id'},
                            'data': {'sdtype': 'categorical'}
                        }
                    },
                },
                'relationships': [
                    {
                        'parent_table_name': 'parent1',
                        'parent_primary_key': 'parent_ID1',
                        'child_table_name': 'child1',
                        'child_foreign_key': 'parent_ID1'
                    },
                    {
                        'parent_table_name': 'parent2',
                        'parent_primary_key': 'parent_ID2',
                        'child_table_name': 'child2',
                        'child_foreign_key': 'parent_ID2'
                    },
                    {
                        'parent_table_name': 'child1',
                        'parent_primary_key': 'child_ID',
                        'child_table_name': 'grandchild',
                        'child_foreign_key': 'child_ID1'
                    },
                    {
                        'parent_table_name': 'child2',
                        'parent_primary_key': 'child_ID',
                        'child_table_name': 'grandchild',
                        'child_foreign_key': 'child_ID1'
                    },
                ]
            })
            synthesizer = HMASynthesizer(metadata)

            # Run
            synthesizer.fit(data)
            samples = synthesizer.sample(scale=1)

            # Assert tables are the same
            assert set(samples) == set(data)

            # Assert columns are the same
            for table_name, table in samples.items():
                assert set(table.columns) == set(data[table_name].columns)

            # Assert data values all exist in the original tables
            for table_name, table in samples.items():
                assert table['data'].isin(data[table_name]['data']).all()

```

When we run the integration test above we get this error:

```
            if (parent_name, child_name) not in added_relationships:
                self._add_foreign_key_columns(
>                   sampled_data[child_name],
                    sampled_data[parent_name],
                    child_name,
                    parent_name
                )
E               KeyError: 'child1'

sdv/sampling/hierarchical_sampler.py:213: KeyError
```

We think the graph traversal algorithm implemented contains a bug. The last child node to be traversed is never sampled by the BaseHierarchicalSampler. When adding relationships in the BaseHierarchicalSampler, the last child (in the provided exampled, child1) is not found in the sampled_data dictionary causing the error.

### Notes
We switched to the latest development version (1.4.1.dev0) as we found a similar error when sampling the provided model with the latest stable version (1.4.0). We observed that In version 1.4.0 the hierarchical sampler was not able to sample more than one level of depth.

--- a/sdv/sampling/hierarchical_sampler.py
+++ b/sdv/sampling/hierarchical_sampler.py
@@ -117,8 +117,9 @@ class BaseHierarchicalSampler():
     def _sample_children(self, table_name, sampled_data):
         """Recursively sample the children of a table.
 
-        This method will loop through the children of a table and sample rows for that child for 
-        every primary key value in the parent.
+        This method will loop through the children of a table and sample rows for that child for
+        every primary key value in the parent. If the child has already been sampled by another
+        parent, this method will skip it.
 
         Args:
             table_name (string):
@@ -202,8 +203,8 @@ class BaseHierarchicalSampler():
             parent_name = relationship['parent_table_name']
             child_name = relationship['child_table_name']
             # When more than one relationship exists between two tables, only the first one
-            # is used to recreate the child tables, so that's the only one that needs to be
-            # added back.
+            # is used to recreate the child tables, so the rest of the foreign key columns
+            # need to be added.
             if (parent_name, child_name) not in added_relationships:
                 self._add_foreign_key_columns(
                     sampled_data[child_name],

--- a/sdv/sampling/hierarchical_sampler.py
+++ b/sdv/sampling/hierarchical_sampler.py
@@ -185,7 +185,7 @@ class BaseHierarchicalSampler():
         """
         sampled_data = {}
 
-        # Sample root tables
+        # DFS to sample roots and then their children
         non_root_parents = set(self.metadata._get_parent_map().keys())
         root_parents = set(self.metadata.tables.keys()) - non_root_parents
         for table in root_parents:
@@ -193,18 +193,14 @@ class BaseHierarchicalSampler():
             synthesizer = self._table_synthesizers[table]
             LOGGER.info(f'Sampling {num_rows} rows from table {table}')
             sampled_data[table] = self._sample_rows(synthesizer, num_rows)
-
-        # Sample rest of the graph
-        for root_table in root_parents:
-            self._sample_children(table_name=root_table, sampled_data=sampled_data)
+            self._sample_children(table_name=table, sampled_data=sampled_data)
 
         added_relationships = set()
         for relationship in self.metadata.relationships:
             parent_name = relationship['parent_table_name']
             child_name = relationship['child_table_name']
             # When more than one relationship exists between two tables, only the first one
-            # is used to recreate the child tables, so the rest of the foreign key columns
-            # need to be added.
+            # is used to recreate the child tables, so the rest can be skipped.
             if (parent_name, child_name) not in added_relationships:
                 self._add_foreign_key_columns(
                     sampled_data[child_name],

--- a/sdv/sampling/hierarchical_sampler.py
+++ b/sdv/sampling/hierarchical_sampler.py
@@ -114,22 +114,18 @@ class BaseHierarchicalSampler():
                 sampled_data[child_name] = pd.concat(
                     [previous, sampled_rows]).reset_index(drop=True)
 
-    def _sample_table(self, table_name, sampled_data):
-        """Recursively sample every table.
+    def _sample_children(self, table_name, sampled_data):
+        """Recursively sample the children of a table.
 
-        Sample top level tables first, then their children, and so on.
+        This method will loop through the children of a table and sample rows for that child for 
+        every primary key value in the parent.
 
         Args:
             table_name (string):
-                Name of the table to sample.
+                Name of the table to sample children for.
             sampled_data (dict):
                 A dictionary mapping table names to sampled tables (pd.DataFrame).
         """
-        for parent_name in self.metadata._get_parent_map()[table_name]:
-            if parent_name not in sampled_data:
-                self._sample_table(table_name=parent_name, sampled_data=sampled_data)
-                return  # Optimization to avoid iterating through the same nodes multiple times
-
         for child_name in self.metadata._get_child_map()[table_name]:
             if child_name not in sampled_data:  # Sample based on only 1 parent
                 for _, row in sampled_data[table_name].iterrows():
@@ -139,7 +135,7 @@ class BaseHierarchicalSampler():
                         parent_row=row,
                         sampled_data=sampled_data
                     )
-                self._sample_table(table_name=child_name, sampled_data=sampled_data)
+                self._sample_children(table_name=child_name, sampled_data=sampled_data)
 
     def _finalize(self, sampled_data):
         """Remove extra columns from sampled tables and apply finishing touches.
@@ -198,8 +194,8 @@ class BaseHierarchicalSampler():
             sampled_data[table] = self._sample_rows(synthesizer, num_rows)
 
         # Sample rest of the graph
-        starting_table = root_parents.pop()  # Start at any root table
-        self._sample_table(table_name=starting_table, sampled_data=sampled_data)
+        for root_table in root_parents:
+            self._sample_children(table_name=root_table, sampled_data=sampled_data)
 
         added_relationships = set()
         for relationship in self.metadata.relationships:

--- a/tests/integration/multi_table/test_hma.py
+++ b/tests/integration/multi_table/test_hma.py
@@ -932,3 +932,131 @@ class TestHMASynthesizer:
         # Assert data values all exist in the original tables
         for table_name, table in samples.items():
             assert table['data'].isin(data[table_name]['data']).all()
+
+    def test_hma_two_lineages_one_grandchild(self):
+        """Test it works on a dataset where one grandchild comes from two lineages.
+
+        Dataset has the shape:
+        r1    r2
+        \\    //
+         c1  c2
+          \\//
+           gc
+        """
+        # Setup
+        root1 = pd.DataFrame(data={
+            'id': [0, 1, 2, 3, 4],
+            'data': [True, False, False, False, True]
+        })
+        root2 = pd.DataFrame(data={
+            'id': [0, 1, 2, 3, 4],
+            'data': [True, False, False, False, True]
+        })
+        child1 = pd.DataFrame(data={
+            'child_ID': ['a', 'b', 'c', 'd', 'e'],
+            'root1_ID': [0, 1, 2, 3, 3],
+            'data': [True, False, False, False, True]
+        })
+        child2 = pd.DataFrame(data={
+            'child_ID': ['a', 'b', 'c', 'd', 'e'],
+            'root2_ID': [0, 1, 2, 3, 4],
+            'data': [True, False, False, False, True]
+        })
+        grandchild = pd.DataFrame(data={
+            'grandchild_ID': ['a', 'b', 'c', 'd', 'e'],
+            'child1_ID': ['a', 'b', 'c', 'd', 'e'],
+            'child2_ID': ['a', 'b', 'c', 'd', 'e'],
+            'data': [True, False, False, False, True]
+        })
+        data = {
+            'root1': root1,
+            'root2': root2,
+            'child1': child1,
+            'child2': child2,
+            'grandchild': grandchild
+        }
+        metadata = MultiTableMetadata.load_from_dict({
+            'tables': {
+                'root1': {
+                    'primary_key': 'id',
+                    'columns': {
+                        'id': {'sdtype': 'id'},
+                        'data': {'sdtype': 'categorical'}
+                    }
+                },
+                'root2': {
+                    'primary_key': 'id',
+                    'columns': {
+                        'id': {'sdtype': 'id'},
+                        'data': {'sdtype': 'categorical'}
+                    }
+                },
+                'child1': {
+                    'primary_key': 'child_ID',
+                    'columns': {
+                        'child_ID': {'sdtype': 'id'},
+                        'root1_ID': {'sdtype': 'id'},
+                        'data': {'sdtype': 'categorical'}
+                    }
+                },
+                'child2': {
+                    'primary_key': 'child_ID',
+                    'columns': {
+                        'child_ID': {'sdtype': 'id'},
+                        'root2_ID': {'sdtype': 'id'},
+                        'data': {'sdtype': 'categorical'}
+                    }
+                },
+                'grandchild': {
+                    'primary_key': 'grandchild_ID',
+                    'columns': {
+                        'grandchild_ID': {'sdtype': 'id'},
+                        'child1_ID': {'sdtype': 'id'},
+                        'child2_ID': {'sdtype': 'id'},
+                        'data': {'sdtype': 'categorical'}
+                    }
+                },
+            },
+            'relationships': [
+                {
+                    'parent_table_name': 'root1',
+                    'parent_primary_key': 'id',
+                    'child_table_name': 'child1',
+                    'child_foreign_key': 'root1_ID'
+                },
+                {
+                    'parent_table_name': 'root2',
+                    'parent_primary_key': 'id',
+                    'child_table_name': 'child2',
+                    'child_foreign_key': 'root2_ID'
+                },
+                {
+                    'parent_table_name': 'child1',
+                    'parent_primary_key': 'child_ID',
+                    'child_table_name': 'grandchild',
+                    'child_foreign_key': 'child1_ID'
+                },
+                {
+                    'parent_table_name': 'child2',
+                    'parent_primary_key': 'child_ID',
+                    'child_table_name': 'grandchild',
+                    'child_foreign_key': 'child2_ID'
+                },
+            ]
+        })
+        synthesizer = HMASynthesizer(metadata)
+
+        # Run
+        synthesizer.fit(data)
+        samples = synthesizer.sample(scale=1)
+
+        # Assert tables are the same
+        assert set(samples) == set(data)
+
+        # Assert columns are the same
+        for table_name, table in samples.items():
+            assert set(table.columns) == set(data[table_name].columns)
+
+        # Assert data values all exist in the original tables
+        for table_name, table in samples.items():
+            assert table['data'].isin(data[table_name]['data']).all()

--- a/tests/unit/sampling/test_hierarchical_sampler.py
+++ b/tests/unit/sampling/test_hierarchical_sampler.py
@@ -1,5 +1,5 @@
 from collections import defaultdict
-from unittest.mock import Mock
+from unittest.mock import Mock, call
 
 import numpy as np
 import pandas as pd
@@ -177,16 +177,13 @@ class TestBaseHierarchicalSampler():
         })
         pd.testing.assert_frame_equal(sampled_data['sessions'], expected_result)
 
-    def test__sample_table(self):
-        """Test sampling a table.
-
-        The ``_sample_table`` method will call sample children and return the sampled data
-        dictionary.
+    def test__sample_children(self):
+        """Test sampling the children of a table.
 
         ``_sample_table`` does not sample the root parents of a graph, only the children.
         """
         # Setup
-        def sample_children(table_name, sampled_data, table_rows):
+        def sample_children(table_name, sampled_data):
             sampled_data['sessions'] = pd.DataFrame({
                 'user_id': [1, 1, 3],
                 'session_id': ['a', 'b', 'c'],
@@ -204,14 +201,10 @@ class TestBaseHierarchicalSampler():
         instance._table_sizes = {'users': 10, 'sessions': 5, 'transactions': 3}
         instance._table_synthesizers = {'users': Mock()}
         instance._sample_children.side_effect = sample_children
-        instance._sample_rows.return_value = pd.DataFrame({
-            'user_id': [1, 2, 3],
-            'name': ['John', 'Doe', 'Johanna']
-        })
 
         # Run
         result = {'users': pd.DataFrame()}
-        BaseHierarchicalSampler._sample_table(
+        BaseHierarchicalSampler._sample_children(
             self=instance,
             table_name='users',
             sampled_data=result
@@ -309,26 +302,34 @@ class TestBaseHierarchicalSampler():
             pd.testing.assert_frame_equal(result_frame, expected_frame)
 
     def test__sample(self):
-        """Test that the ``_sample_table`` is called for root tables."""
-        # Setup
-        expected_sample = {
-            'users': pd.DataFrame({
-                'user_id': [1, 2, 3],
-                'name': ['John', 'Doe', 'Johanna']
-            }),
-            'sessions': pd.DataFrame({
-                'user_id': [1, 1, 3],
-                'session_id': ['a', 'b', 'c'],
-                'os': ['windows', 'linux', 'mac'],
-                'country': ['us', 'us', 'es']
-            }),
-            'transactions': pd.DataFrame(dtype='Int64')
-        }
+        """Test that the whole dataset is sampled.
 
-        def _sample_table(table_name, sampled_data):
-            sampled_data['users'] = expected_sample['users']
-            sampled_data['sessions'] = expected_sample['sessions']
-            sampled_data['transactions'] = expected_sample['transactions']
+        Sampling has the following steps:
+        1. The root tables should be sampled first.
+        2. Then the lineage for each root is sampled by calling ``_sample_children``.
+        3. Any missing parent-child relationships are added using ``_add_foreign_key_columns``.
+        4. All extra columns are dropped by calling ``_finalize``.
+        """
+        # Setup
+        users = pd.DataFrame({
+            'id': [1, 2, 3],
+            'name': ['John', 'Doe', 'Johanna']
+        })
+        sessions = pd.DataFrame({
+            'user_id': [1, 1, 3],
+            'session_id': ['a', 'b', 'c'],
+            'os': ['windows', 'linux', 'mac'],
+            'country': ['us', 'us', 'es']
+        })
+        transactions = pd.DataFrame({
+            'user_id': [1, 2, 3],
+            'transaction_id': [1, 2, 3],
+            'transaction_amount': [100, 1000, 200]
+        })
+
+        def _sample_children_dummy(table_name, sampled_data):
+            sampled_data['sessions'] = sessions
+            sampled_data['transactions'] = transactions
 
         instance = Mock()
         instance._table_sizes = {
@@ -341,32 +342,67 @@ class TestBaseHierarchicalSampler():
                 'parent_table_name': 'users',
                 'parent_primary_key': 'id',
                 'child_table_name': 'sessions',
-                'child_foreign_key': 'id'
+                'child_foreign_key': 'user_id'
+            },
+            {
+                'parent_table_name': 'users',
+                'parent_primary_key': 'id',
+                'child_table_name': 'transactions',
+                'child_foreign_key': 'user_id'
             }
         ]
         users_synthesizer = Mock()
         instance._table_synthesizers = defaultdict(Mock, {'users': users_synthesizer})
         instance.metadata._get_parent_map.return_value = {
             'sessions': ['users'],
-            'transactions': ['sessions']
+            'transactions': ['users']
         }
         instance.metadata.tables = {
             'users': Mock(),
             'sessions': Mock(),
             'transactions': Mock(),
         }
-        instance._sample_table.side_effect = _sample_table
+        instance._sample_rows.return_value = users
+        instance._sample_children.side_effect = _sample_children_dummy
 
         # Run
         result = BaseHierarchicalSampler._sample(instance)
 
         # Assert
+        expected_sample = {
+            'users': DataFrameMatcher(pd.DataFrame({
+                'id': [1, 2, 3],
+                'name': ['John', 'Doe', 'Johanna']
+            })),
+            'sessions': DataFrameMatcher(pd.DataFrame({
+                'user_id': [1, 1, 3],
+                'session_id': ['a', 'b', 'c'],
+                'os': ['windows', 'linux', 'mac'],
+                'country': ['us', 'us', 'es']
+            })),
+            'transactions': DataFrameMatcher(pd.DataFrame({
+                'user_id': [1, 2, 3],
+                'transaction_id': [1, 2, 3],
+                'transaction_amount': [100, 1000, 200]
+            }))
+        }
         assert result == instance._finalize.return_value
-        instance._sample_table.assert_called_once_with(table_name='users',
-                                                       sampled_data=expected_sample)
-        instance._add_foreign_key_columns.assert_called_once_with(
-            DataFrameMatcher(expected_sample['sessions']),
-            DataFrameMatcher(expected_sample['users']),
-            'sessions',
-            'users')
+        instance._sample_children.assert_called_once_with(
+            table_name='users',
+            sampled_data=expected_sample
+        )
+        instance._add_foreign_key_columns.assert_has_calls([
+            call(
+                expected_sample['sessions'],
+                expected_sample['users'],
+                'sessions',
+                'users'
+            ),
+            call(
+                expected_sample['transactions'],
+                expected_sample['users'],
+                'transactions',
+                'users'
+            )
+        ])
         instance._finalize.assert_called_once_with(expected_sample)

